### ISSUE TITLE
comptest takes 5mins to run on OS x

### ISSUE THREAD

**Issue by charliegracie** [eclipse/omr#1811]
I have been trying to improve the Travis build/test times and I noticed that OS X is still a lot longer than Linux x86.  One of the biggest differences is the time spent in comptest.  On Linux x86 it takes ~4s but on OS X it takes ~350s.  It is actually 2 of the subtests which take most of the time.

Linux x86:
1 test from CommoningDeathTest (105 ms total)
13 tests from ILValidatorDeathTest/IllformedTrees (1305 ms total)

OS X:
1 test from CommoningDeathTest (18580 ms total)
13 tests from ILValidatorDeathTest/IllformedTrees (324226 ms total)

**Comment by charliegracie** [comment]
@Leonardo2718 this seems really strange to me.  When I was working on SOM++ though I noticed that the std::map data structure was orders of magnitudes of slower on OS X than it was on Linux.  The trivial use case in SOM++ was adding a 2X throughput loss for the entire benchmark run so I could believe something like that could be causing this issue. FYI @mstoodle and @0xdaryl

**Comment by Leonardo2718** [comment]
Without profiling information, it will be difficult to pin-point what exactly is consuming so much time.

However, my initial thought is that it probably has more to do with the Death Test implementation in gtest rather than the tests themselves. Those death tests always run slower than any other test.

If this is correct, I don't think there's a lot that can be done, though it might be worth researching if someone else has also run into perf problems with gtest on OSX and found a way around it.

**Comment by charliegracie** [comment]
I finally got my OS X machine up and running.  If I run comptest by itself it only takes a few seconds so the issue was not quickly reproducible on my machine.  I will run it under ctest and see if that makes any difference.  If it does not then I will check my system configuration to see what if any differences there are.

**Comment by charliegracie** [comment]
I am using AppleClang 9.0.0.9000038 and the Travis builds are using AppleClang 7.3.0.7030031.

**Comment by charliegracie** [comment]
I tried a build that used [XCode 9](https://travis-ci.org/charliegracie/omr/jobs/288795551).  The time to run the comptest was not improved.

This was the compiler version in that test: AppleClang 9.0.0.9000037

**Comment by charliegracie** [comment]
I got a co-worker who has OS X 10.11 to try and the time was also fast on his machine.

My next step will be to see if I can use "Instruments" on the command line to gather performance data.

**Comment by charliegracie** [comment]
I was able to get profiling information from the travis build using `sample`.  Most of the time spent in the *Death*Tests is actually in `read()`.  Here is a sample stack trace:
```
41784 IllformedTrees_FailCompilation_Test::TestBody()  (in comptest) + 670  [0x105ed7bae]  ILValidatorTest.cpp:37
      + 41784 testing::internal::ForkingDeathTest::Wait()  (in comptest) + 72  [0x105f798d8]
      +   41784 testing::internal::DeathTestImpl::ReadAndInterpretStatusByte()  (in comptest) + 54  [0x105f77696]
      +     41784 testing::internal::posix::Read(int, void*, unsigned int)  (in comptest) + 33  [0x105f90901]
      +       41784 read  (in libsystem_kernel.dylib) + 10  [0x7fff8b4a2362]
```
For some reason the `read` API on the travis OS X machines is significantly slower than any of the local machines I have been able to use.  We may want to disable these tests on OS X for the time being.

I actually do not like that we are testing that the compiler crashes (asserts) on bad IL.  It would be better for me if that caused a failure with a non-zero return code instead of taking the process down.  @mstoodle  thoughts?

**Comment by mstoodle** [comment]
@charliegracie I believe it's validating that an assertion was triggered, which I think is an easier to maintain model than creating a new return code for every possible kind of check.

@Leonardo2718 is my understanding correct?

**Comment by Leonardo2718** [comment]
So, by default, a validation failure will take down the process in DEBUG and PROD_WTH_ASSUMES builds. In a PROD build, it will throw an exception that should only abort compilation (likely resulting in a non-zero return code). Because these tests are run using a PROD_WITH_ASSUMES build, death tests are the only way of catching failures.

Death tests do have the advantage of being able to check the failure message. In these tests, this feature is only used to unsure that a validation failure has occurred, but it's so granular so as to detect which opcode caused the failure.

I am not particularly attached to this behaviour, so I would be okay with just aborting compilation and returning an appropriate error code. The error code should be unique to validation failures though, so that tests can verify the failure is indeed an IL validation error.

Another alternative would be let the compiler throw an exception and pass it through to the test (i.e. not catching it and turning it into a return code). Then an `ASSERT_THROW` could be used to catch these failures.

Perhaps this discussion deserves to have its own issue opened, since it's more general than the main topic of this issue (`comptest` taking too long).

**Comment by mgaudet** [comment]
I agree w/ Leo -- we can tackle the mechanics of the validator tests in another issue. I suspect that given this seems to be getting worse, in the mean time, we should probably change `.Travis.yml` to disable the DeathTests on OS/X.
